File: Advanced_Program_BFSMaze/referenceCode/DisjointSet.cpp
// DisjointSet implementation using union by size and path compression
// By Mary Elaine Califf and Jake Clouse

#include "DisjointSet.h"
#include <iostream>

DisjointSet::DisjointSet(int numObjects){
    //to do
    numValues = numObjects;
    theArray = std::vector<int>(numObjects, -1);
}

//recursive method to find the item -- does path compression on the way out of the recursion
int DisjointSet::find(int objectIndex){
    // to do -- see assignment instructions for details
    if( theArray[ objectIndex ] < 0 )
        return objectIndex;
    else
        return theArray[ objectIndex ] = find( theArray[ objectIndex ] );
}

bool DisjointSet::doUnion(int objIndex1, int objIndex2){
    // to do -- see assignment instructions for details
    // printArrayValues(std::cout);

    bool foundNegative = false;
    int root1 = find(objIndex1);
    int root2 = find(objIndex2);
    if (theArray[root2]>=theArray[root1]){
       theArray[root1] += theArray[root2];
       theArray[root2] = root1;
    }
    else{
        theArray[root2] += theArray[root1];
        theArray[root1] = root2;
    }
    for(int i = 0; i < theArray.size(); i++){
        if(theArray[i] < 0){
            if(!foundNegative){
                foundNegative = true;
            }
            else{
                return false;
            }
        }
    }
    return true;
}

void DisjointSet::printArrayValues(std::ostream &outputStream){
    for (int i = 0; i < numValues; i++)
    {
        outputStream << theArray[i] << " ";
    }
    outputStream << std::endl;
}

File: Advanced_Program_BFSMaze/DisjointSet.go
package main

import "fmt"

type DisjointSet struct {
	numElements int
	theArray   []int
}

func NewDisjointSet(size int) *DisjointSet {
	ds := &DisjointSet{
		numElements: size,
		theArray:   make([]int, size),
	}
	for i:=0 ; i < size; i++ {
		ds.theArray[i] = -1
	}
	return ds
}

func (ds *DisjointSet) Find(index int) int {
	if(ds.theArray[index] < 0){
		return index
	} else{
		ds.theArray[ index ] = ds.Find( ds.theArray[ index ] )
		return ds.theArray[ index ]
	}
}

func (ds *DisjointSet) Union(index1 int, index2 int) bool {
	var foundNegative bool = false
	var root1 int = ds.Find(index1)
	var root2 int = ds.Find(index2)
	if(ds.theArray[root2]>=ds.theArray[root1]){
		ds.theArray[root1] += ds.theArray[root2]
		ds.theArray[root2] = root1
	} else {
		ds.theArray[root2] += ds.theArray[root1]
		ds.theArray[root1] = root2	
	}
	for i:=0; i<ds.numElements; i++{
		if (ds.theArray[i] < 0){
			if(!foundNegative){
				foundNegative = true
			}else{
				return false
			}
		}
	}
	return true
}
func (ds *DisjointSet) PrintArrayValues() {
	for i:= 0; i<ds.numElements; i++ {
		fmt.Print(ds.theArray[i]," ")
	}
	fmt.Print("\n")
}
